pkg/ipcache/ipcache.go

```diff
@@ -434,6 +435,41 @@ func (ipc *IPCache) DumpToListener(listener IPIdentityMappingListener) {
 	ipc.RUnlock()
 }
 
+// UpsertIdentity is currently unused. The idea is that this API will interface
+// with the IPCache.metadata to associate the identity with the prefix, causing
+// the identity determination to be overruled based on the source here, but
+// still merge all of the other metadata from the IPCache.metadata map like the
+// encrypt key.
+func (ipc *IPCache) UpsertIdentity(addr netip.Addr, hostIP netip.Addr, id Identity, src source.Source, resource types.ResourceID, aux ...IPMetadata) {
+	// We expose hostIP, id to ensure the caller provides it, but internally
+	// these are just another piece of aux data associated with the prefix.
+	// TODO: Add them into the slice.
+	//aux = append(aux, IPMetadata(hostIP))
+	//aux = append(aux, IPMetadata(id))
+
+	prefix := netip.PrefixFrom(addr, addr.BitLen())
+	ipc.metadata.Lock()
+	ipc.metadata.upsertLocked(prefix, source.Generated, resource, aux...)
+	ipc.metadata.Unlock()
+	ipc.metadata.enqueuePrefixUpdates(prefix)
+	ipc.TriggerLabelInjection()
+
+	// TODO: It would be nice to make this play nice with InjectLabels()
+	// to ensure proper identity allocation/release when different
+	// identities are associated with a particular prefix.
+	// Could we get away with making this asynchronous too?
+	// Also store/merge info like encrypt keys, k8sMeta in ipc.metadata?
```

Comment:
Why is the host key a separate parameter here? I also have a local branch handles encrypt key and host ip as part of `InjectLabels` (based on your previous work), where treat it as an IPMetadata argument.

The tricky thing with all external metadata (encrypt key/host IP/k8sMeta) is that they could still get overwritten by someone using the legacy API for a prefix entry that was created via the async API. 

One solution would be to have the legacy API check if there is a host ip or encrypt key in the metadata, and if so, not overwrite it. But that has lock ordering complications and I'm not sure it will be always correct.

Another solution here might be to require "never mix the old and new APIs for the same prefix", but it's hard to validate and I'm not even sure we can guarantee that at the moment . The node manager is a known offender here, but it can be ported relatively easily I think. I'm less certain about other parts of the code (especially when it comes to CEPs).

---

I'm also wondering if it makes sense to split up `K8sMetadata` into three or more separate `IPMetadata` keys. Probably not initially to make the transformation easier, but I always found it a bit weird that `K8sMetadata` contains both named ports (significant for policy) and the pod name/namespace (which is not significant for policy and just there for Hubble)

Reply:
The goal of the parameter was more just to make it explicit that this is a requirement, so we don't inadvertently create an ipcache entry with the hostIP missing. But maybe that's a paranoid requirement that should be better covered by just testing that the entries are populated the way they should be. I'd be OK injecting them as `IPMetadata`.

> The tricky thing with all external metadata (encrypt key/host IP/k8sMeta) is that they could still get overwritten by someone using the legacy API for a prefix entry that was created via the async API.
>
> One solution would be to have the legacy API check if there is a host ip or encrypt key in the metadata, and if so, not overwrite it. But that has lock ordering complications and I'm not sure it will be always correct.
>
> Another solution here might be to require "never mix the old and new APIs for the same prefix", but it's hard to validate and I'm not even sure we can guarantee that at the moment . The node manager is a known offender here, but it can be ported relatively easily I think. I'm less certain about other parts of the code (especially when it comes to CEPs).

My initial thought was the latter. Thinking a bit more, it's pretty nuanced though. That said, for pod metadata I think the whole idea is that the pod information is taking precedence over any other information, so as long as we convert all of the inputs for pod metadata (encrypt/k8smeta/CEP(or equivalent kvstore)) then it should be OK?

I'm open to the idea of grabbing the metadata lock briefly to check, and discarding the Upsert in that case with a loud warning. Metadata lock should in general be held relatively briefly, not held while updating policy etc.

Reply:
> I'm also wondering if it makes sense to split up K8sMetadata into three or more separate IPMetadata keys. Probably not initially to make the transformation easier, but I always found it a bit weird that K8sMetadata contains both named ports (significant for policy) and the pod name/namespace (which is not significant for policy and just there for Hubble)

:+1: SGTM.